### DISCUSSION TITLE
Data science and kart: incorporating non-geospatial data and long-term archiving for repositories

### DISCUSSION THREAD

**Discussion by tcwilkinson** [koordinates/kart#974] [Q&A]
I am currently in process of learning kart, as have been searching for a good spatial and database version control system for a while. 

My use cases are in archaeological data science. I already use git to track spatial files mostly in [original format] GeoJSON and tabular data in CSV, with accompanying explanatory info in plain text or markdown. 

It's important that the working archives are in plain text as much as possible to facilitate long-term archiving of data, but also because it reduces the chances of format orphaning.  (Archaeological projects often take years if not decades to go from field to publication, yet there are few data standards, so making things easy to navigate is important).  

Git helps, but tracking changes with spatial and tabular data is still clunky (not surprising, it was designed for line-based code). So far, I like what I see in kart for spatial data... and would like to adopt it for some projects. But have some questions about the direction of travel and how it would fit with data science workflows, so wanted to open a discussion here for both kart developers and wider users.

1. How easy is it to incorporate non-spatial data into the same repository? For example, text files which should be in version-sync with the data, to help readers to interpret and re-use the data schemas in use, or the history of the data production. (I see the attachment feature, https://github.com/koordinates/kart/issues/583 , but it currently seems unfinished).
3. How human-readable are kart repositories long-term? If kart was no longer developed, would the next generation of data scientists be able to decode easily and re-use the kart-managed part of the repository (either the history or the latest version)?

**Comment by rcoup** [comment]
Hi Toby,

Thanks for reaching out

> How easy is it to incorporate non-spatial data into the same repository? For example, text files which should be in version-sync with the data, to help readers to interpret and re-use the data schemas in use, or the history of the data production. (I see the attachment feature, https://github.com/koordinates/kart/issues/583 , but it currently seems unfinished).

It'll definitely happen, but it's not on the do-next list at the moment.

> How human-readable are kart repositories long-term? If kart was no longer developed, would the next generation of data scientists be able to decode easily and re-use the kart-managed part of the repository (either the history or the latest version)?

Kart is built on Git which has a strong history of backwards compatibility and legacy platform support, and we [document the format and internals](https://docs.kartproject.org/en/latest/pages/development/table_v3.html) of how we store features/etc within Git objects. So while Kart repositories aren't human readable, doing a one-off conversion/extract (eg: to GeoJSON) _without any Kart code running_ would not be a particularly complex task for a developer.

**Reply by bheberlein** [reply]
I can think of at least a few use cases where tracking text files can be very useful:
- Source code, tests & scripts related to a project or analysis
- Project configuration or metadata files (e.g. JSON, YML, TOML files)
- Documentation & examples (e.g. Markdown, Jupyter notebook)

This is relevant for research work, e.g. geospatial modeling & algorithm development.

These use cases also overlap heavily with Git & software development workflows, so it's worth questioning if it's the role of Kart to handle these things. There are a lot of nice tools out there already that cater to software development & code management (e.g. VS Code, PyCharm, Git Kraken) and it's hard to imagine that Kart will end up being the better option for managing text-based repositories unless it can somehow interface with those tools (even if geospatial data is not exposed). However I do think that some way to integrate version control between text/code & geospatial data would be hugely valuable for research, modeling/analysis & archival work, and probably others.

I had thought about using Git submodules to place my Kart repository inside a Git repository, but it doesn't work as the external repo doesn't understand the Kart index. Is there a different strategy that can achieve something like this with existing tools?

I don't know what's the right approach, but I would be very excited to see some discussion or movement on this (or get some recommendations). Currently I'm just tinkering & learning the ropes, but trying to decide if I should fully adopt Kart for my work. How to link version control of source code, documentation & data is a big open question for me that makes me hesitant to fully adopt.

I have also looked at [DVC](https://dvc.org), which is targeted toward integrating data version control with Git for modeling (primarily ML). It may be nice for binary or image data, or nonconventional formats, which could be geospatial in nature, but it is not nearly as well suited to GIS (esp. vector) data compared to Kart.

**Comment by olsen232** [comment]
Hi Brendan,

I wish it were not so but unfortunately I will have to quote myself from a year ago:

> You are right - attachment support is unfinished, and has been for some time.

Here is the comment where this quote is sourced and with it an explanation for how one _can_ use Kart to store non-geospatial attachments (text files, PDFs) in a version controlled way.

https://github.com/koordinates/kart/issues/583#issuecomment-1995818262

However as you can see this is pretty bare-bones and I can't recommend it for anything too complicated. 
For instance, if you want to have version controlled geospatial data, and a version controlled source code repo for tests and scripts related to the project, at this stage my personal preference would be to store them in two separate repos, Kart for geospatial and regular git for source code.
In the source code repo you could store a file or script which references a particular Kart commit, as a way to keep them in sync eg.
```
repos
├── my-kart-data
│   ├── .kart
│   └── my-kart-data.gpkg
└── my-source-code
    ├── .git
    └── sync-kart-repo.sh
```

Where the contents of `sync-kart-repo.sh` is something like `kart -C ~/repos/my-kart-data checkout 87dde8a3`
but you check in a new version of that file from time to time, similar to how when you update a submodule you need to make a commit in the outer module which references the new version of the submodule.
